### PR TITLE
Add system schema store

### PR DESCRIPTION
**Pull request by Linary**
Change-Id: I2feabace71f266d6be6c2094c30cd19f74ab7f7d

### CODE REVIEW

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/SystemSchemaStore.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+package com.baidu.hugegraph.backend.store;
+
+import java.util.HashMap;
+
+import com.baidu.hugegraph.backend.id.Id;
+import com.baidu.hugegraph.schema.SchemaElement;
+
+public class SystemSchemaStore {
+
+    private final HashMap<Id, SchemaElement> store;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/SystemSchemaStore.java:29]: use arraymap for perf

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/SchemaTransaction.java [hunk 2]
```diff
@@ -294,6 +301,11 @@ protected void addSchema(SchemaElement schema) {
     protected <T extends SchemaElement> T getSchema(HugeType type, Id id) {
         LOG.debug("SchemaTransaction get {} by id '{}'",
                   type.readableName(), id);
+        // System schema just get from SystemSchemaStore in memory
+        if (id.asLong() < 0) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/SchemaTransaction.java:305]: 0L

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/SchemaTransaction.java [hunk 3]
```diff
@@ -294,6 +301,11 @@ protected void addSchema(SchemaElement schema) {
...
+            return this.store().systemSchemaStore().get(id);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/SchemaTransaction.java:306]: define field var for this.store().systemSchemaStore()

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/SchemaTransaction.java [hunk 4]
```diff
@@ -315,6 +327,8 @@ protected void addSchema(SchemaElement schema) {
                                                     String name) {
         LOG.debug("SchemaTransaction get {} by name '{}'",
                   type.readableName(), name);
+        // Assume that system schema will not get by name
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/SchemaTransaction.java:330]: seems can't assume that

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/SchemaTransaction.java [hunk 5]
```diff
@@ -350,6 +364,12 @@ protected void addSchema(SchemaElement schema) {
     protected void removeSchema(SchemaElement schema) {
         LOG.debug("SchemaTransaction remove {} by id '{}'",
                   schema.type(), schema.id());
+        // System schema just remove from SystemSchemaStore in memory
+        if (schema.longId() < 0) {
+            this.store().systemSchemaStore().remove(schema.id());
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/SchemaTransaction.java:369]: ditto

**Review by javeme — DISMISSED** [review]

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/SystemSchemaStore.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+    private static final int SYSTEM_SCHEMA_MAX_NUMS = 128;
+
+    private SchemaElement[] storeByIds;
+    private final Map<String, SchemaElement> storeByNames;
+
+    public SystemSchemaStore() {
+        CollectionFactory factory = new CollectionFactory(CollectionType.EC);
+        this.storeByIds = new SchemaElement[SYSTEM_SCHEMA_MAX_NUMS];
+        this.storeByNames = factory.newMap(SYSTEM_SCHEMA_MAX_NUMS);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/SystemSchemaStore.java:43]: CollectionFactory.newMap(CollectionType.EC,SYSTEM_SCHEMA_MAX_NUMS);

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/SystemSchemaStore.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+    @SuppressWarnings("unchecked")
+    public <T extends SchemaElement> T get(String name) {
+        return (T) this.storeByNames.get(name);
+    }
+
+    private void expandCapacity() {
+        int newLength = this.storeByIds.length << 1;
+        SchemaElement[] newStoreByIds = new SchemaElement[newLength];
+        System.arraycopy(this.storeByIds, 0, newStoreByIds, 0,
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/SystemSchemaStore.java:73]: not covered by tests

File: hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBTables.java [hunk 8]
```diff
@@ -34,9 +34,39 @@
...
+        }
+
+        public void writeVersion(Session session, String version) {
+            byte[] key = new byte[]{HugeKeys.VERSION.code()};
+            byte[] value = StringEncoding.encode(version);
+            session.put(this.table(), key, value);
+            try {
+                session.commit();
+            } catch (Exception e) {
+                session.rollback();
```
Comment (javeme) [inline hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBTables.java:56]: rethrow after rollback()

**Comment by Linary** [comment]
To updated

File: hugegraph-hbase/src/main/java/com/baidu/hugegraph/backend/store/hbase/HbaseStore.java [hunk 9]
```diff
@@ -540,5 +540,45 @@ public long getCounter(HugeType type) {
...
+            List<String> tableNames = super.tableNames();
+            tableNames.add(this.meta.table());
+            return tableNames;
+        }
+
+        @Override
+        public void init() {
+            super.init();
+            super.checkOpened();
+            Session session = super.getSession();
```
Comment (javeme) [inline hugegraph-hbase/src/main/java/com/baidu/hugegraph/backend/store/hbase/HbaseStore.java:571]: can use `Session session = super.sessions.session();`

File: hugegraph-hbase/src/main/java/com/baidu/hugegraph/backend/store/hbase/HbaseStore.java [hunk 10]
```diff
@@ -540,5 +540,45 @@ public long getCounter(HugeType type) {
...
+            String driverVersion = this.provider().driverVersion();
+            this.meta.writeVersion(session, driverVersion);
+            LOG.info("Write down the backend version: {}", driverVersion);
+        }
+
+        @Override
+        public String storedVersion() {
+            super.checkOpened();
+            Session session = super.getSession();
```
Comment (javeme) [inline hugegraph-hbase/src/main/java/com/baidu/hugegraph/backend/store/hbase/HbaseStore.java:580]: ditto

File: hugegraph-hbase/src/main/java/com/baidu/hugegraph/backend/store/hbase/HbaseStore.java [hunk 11]
```diff
@@ -540,5 +540,45 @@ public long getCounter(HugeType type) {
...
+
+        @Override
+        public void init() {
+            super.init();
+            super.checkOpened();
+            Session session = super.getSession();
+            String driverVersion = this.provider().driverVersion();
+            this.meta.writeVersion(session, driverVersion);
+            LOG.info("Write down the backend version: {}", driverVersion);
+        }
```
Comment (javeme) [inline hugegraph-hbase/src/main/java/com/baidu/hugegraph/backend/store/hbase/HbaseStore.java:575]: also overwrite truncate() and let it truncate other tables except the meta table

File: hugegraph-cassandra/src/main/java/com/baidu/hugegraph/backend/store/cassandra/CassandraStore.java [hunk 12]
```diff
@@ -670,5 +670,45 @@ public long getCounter(HugeType type) {
...
+            String driverVersion = this.provider().driverVersion();
+            this.meta.writeVersion(session, driverVersion);
+            LOG.info("Write down the backend version: {}", driverVersion);
+        }
+
+        @Override
+        public String storedVersion() {
+            this.checkOpened();
+            CassandraSessionPool.Session session = this.getSession();
+            return this.meta.readVersion(session);
```
Comment (javeme) [inline hugegraph-cassandra/src/main/java/com/baidu/hugegraph/backend/store/cassandra/CassandraStore.java:704]: seems server-info can also be stored into meta table

**Pull request closed without merging**